eng/_util/cmd/generate-codeql-yml/generate-codeql-yml.go

```
@@ -0,0 +1,95 @@
+
+package main
+
+import (
+	"bufio"
+	"fmt"
+	"log"
+	"maps"
+	"os"
+	"path/filepath"
+	"regexp"
+	"slices"
+	"strings"
+)
+
+const (
+	patchDir        = "patches"
+	goSubmodulePath = "go"
+	outputFile      = "CodeQL.yml"
+	yamlHeader      = "# This file is auto-generated by generate-codeql-yml"
+	libraryHeader   = "path_classifiers:\n  library:\n    # Classify the upstream go submodule as a 3rd-party library\n    - " + goSubmodulePath + "\n    # Except for these Go source files which we patch"
+)
+
+var (
+	diffRegex      = regexp.MustCompile(`^diff --git a/(\S+)`)
+	goFileSuffix   = ".go"
+	testFileSuffix = "_test.go"
+)
+
+func main() {
+	pathsToExclude := make(map[string]bool)
+
+	err := filepath.Walk(patchDir, func(path string, info os.FileInfo, err error) error {
+		if err != nil || info.IsDir() || !strings.HasSuffix(path, ".patch") {
+			return nil
+		}
+		file, err := os.Open(path)
+		if err != nil {
+			return err
+		}
+		defer file.Close()
+
+		scanner := bufio.NewScanner(file)
+		for scanner.Scan() {
+			line := scanner.Text()
+			matches := diffRegex.FindStringSubmatch(line)
+			if len(matches) != 2 {
+				continue
+			}
+			filePath := matches[1]
+
+			// Apply exclusion filters
+			if !strings.HasSuffix(filePath, goFileSuffix) {
+				continue
+			}
+			// Skip test files and testdata directories
+			if strings.HasSuffix(filePath, testFileSuffix) {
+				continue
+			}
+			if strings.Contains(filePath, "/testdata/") {
+				continue
+			}
+			// We already run CodeQL on our vendored dependencies
+			if strings.HasPrefix(filePath, "src/vendor/") || strings.HasPrefix(filePath, "src/cmd/vendor/") {
+				continue
+			}
+			pathsToExclude[filePath] = true
+		}
+		return scanner.Err()
+	})
+
+	if err != nil {
+		log.Fatalf("Error walking patches directory: %v", err)
+	}
+
+	// Output YAML
+	f, err := os.Create(outputFile)
+	if err != nil {
+		log.Fatalf("Failed to create %s: %v", outputFile, err)
+	}
+	defer f.Close()
```

Don't ignore error from closing a written-to file, only a read file.